src/Common/src/Interop/Windows/kernel32/Interop.MoveFileEx.cs

```diff
@@ -10,17 +10,35 @@ internal partial class Interop
 {
     internal partial class Kernel32
     {
+        const uint MOVEFILE_REPLACE_EXISTING = 0x01;
+        const uint MOVEFILE_COPY_ALLOWED = 0x02;
+        const uint MOVEFILE_WRITE_THROUGH = 0x08;
+
         /// <summary>
         /// WARNING: This method does not implicitly handle long paths. Use MoveFile.
         /// </summary>
         [DllImport(Libraries.Kernel32, EntryPoint = "MoveFileExW", SetLastError = true, CharSet = CharSet.Unicode, BestFitMapping = false)]
         private static extern bool MoveFileExPrivate(string src, string dst, uint flags);
 
-        internal static bool MoveFile(string src, string dst)
+        /// <summary>
+        /// Moves a file or directory, optionally overwriting existing destination file. NOTE: overwrite must be false for directories.
+        /// </summary>
+        /// <param name="src">Source file or directory</param>
+        /// <param name="dst">Destination file or directory</param>
+        /// <param name="overwrite">True to overwrite existing destination file. NOTE: must pass false for directories as overwrite of directories is not supported.</param>
+        /// <returns></returns>
+        internal static bool MoveFile(string src, string dst, bool overwrite)
         {
             src = PathInternal.EnsureExtendedPrefixIfNeeded(src);
             dst = PathInternal.EnsureExtendedPrefixIfNeeded(dst);
-            return MoveFileExPrivate(src, dst, 2 /* MOVEFILE_COPY_ALLOWED */);
+
+            uint flags = MOVEFILE_WRITE_THROUGH | MOVEFILE_COPY_ALLOWED;
```

Comment:
Why are you adding MOVEFILE_WRITE_THROUGH when it wasn't previously being used in the non-overwrite case?

Reply:
Specifically requested to be set by @JeremyKuhne [here](https://github.com/dotnet/corefx/issues/31511#issuecomment-409686200)

Reply:
> Specifically requested to be set by @JeremyKuhne

@JeremyKuhne, why?  Has there been concern expressed over the current behavior, which presumably has been this way since the beginning of .NET?

Regardless, even if we decide to add the flag, I think it'd be good to be a separate PR, separate from adding new overloads.

Reply:
> why? Has there been concern expressed over the current behavior, which presumably has been this way since the beginning of .NET?

Because it isn't consistent with Unix and isn't deterministic. https://github.com/dotnet/corefx/pull/33054#issuecomment-433519540

Reply:
Please do it as a separate PR such that this one can be specific to the additional overloads and such that discussion focused on whether the flag should be added is isolated.

Reply:
Easy enough. Will remove from this and open as separate PR.

For reference, NetFX [appears to use `MoveFile`](https://referencesource.microsoft.com/#mscorlib/microsoft/win32/win32native.cs,1468) instead of `MoveFileExW`. Not able to find any documentation at the moment on its write-through behavior.

Reply:
> Easy enough. Will remove from this and open as separate PR.

Thanks.

> Because it isn't consistent with Unix and isn't deterministic. 

Are we going to change everything we do with FileStream to always pass FileOptions.WriteThrough?

Reply:
New PR #33078

Reply:
> For reference, NetFX appears to use MoveFile instead of MoveFileExW. Not able to find any documentation at the moment on its write-through behavior.

`MoveFile` just passes `MOVEFILE_COPY_ALLOWED`.

> Are we going to change everything we do with FileStream to always pass FileOptions.WriteThrough?

I don't think they have to match. Calling Copy/Move doesn't have the same level of expectations as FileStream in my mind. You can continue to write data to the cache- attempting to use an old location doesn't work if it hasn't completed.